internal/resolve/catalog.go

```diff
@@ -190,6 +177,46 @@ func (r *CatalogResolver) Resolve(ctx context.Context, ext *ocv1alpha1.ClusterEx
 	return resolvedBundle, resolvedBundleVersion, priorDeprecation, nil
 }
 
+type resolutionError struct {
+	PackageName     string
+	Version         string
+	Channels        []string
+	InstalledBundle *ocv1alpha1.BundleMetadata
+	ResolvedBundles []foundBundle
+}
+
+func (rei resolutionError) Error() string {
+	var sb strings.Builder
+	if rei.InstalledBundle != nil {
+		sb.WriteString(fmt.Sprintf("error upgrading from currently installed version %q: ", rei.InstalledBundle.Version))
+	}
+
+	if len(rei.ResolvedBundles) > 1 {
+		sb.WriteString(fmt.Sprintf("found bundles for package %q ", rei.PackageName))
+	} else {
+		sb.WriteString(fmt.Sprintf("no bundles found for package %q ", rei.PackageName))
+	}
+
+	if rei.Version != "" {
+		sb.WriteString(fmt.Sprintf("matching version %q ", rei.Version))
+	}
+
+	if len(rei.Channels) > 0 {
+		sb.WriteString(fmt.Sprintf("in channels %v", rei.Channels))
+	}
+
+	matchedCatalogs := []string{}
+	for _, r := range rei.ResolvedBundles {
+		matchedCatalogs = append(matchedCatalogs, r.catalog)
+	}
+	slices.Sort(matchedCatalogs) // sort for consistent error message
+	if len(matchedCatalogs) > 1 {
+		sb.WriteString(fmt.Sprintf("in multiple catalogs with the same priority %v", matchedCatalogs))
```

Comment:
Not strictly necessary, but following the pattern and making it easier for additional context to be added later:
```suggestion
		sb.WriteString(fmt.Sprintf("in multiple catalogs with the same priority %v ", matchedCatalogs))
```

Reply:
Should be updated in the latest push